File: src/mcp/endpoints/memory.py
"""
Memory REST Endpoints

Memory-related HTTP endpoints for Desktop frontend.
"""

import base64
import io
import logging
from typing import List

from starlette.requests import Request
from starlette.responses import JSONResponse

logger = logging.getLogger("niko-gateway")


async def memory_search_endpoint(request: Request):
    """Memory search REST endpoint."""
    from src.mcp.gateway import memory_search

    body = await request.json()
    result = await memory_search(
        query=body.get("query", ""),
        layer=body.get("layer"),
        dimensions=body.get("dimensions"),
        entity_id=body.get("entity_id"),
        at_time=body.get("at_time"),
        limit=body.get("limit", 10),
    )
    return JSONResponse(result)


async def memory_add_endpoint(request: Request):
    """Memory add REST endpoint."""
    from src.mcp.gateway import memory_add

    body = await request.json()
    result = await memory_add(
        content=body.get("content", ""),
        layer=body.get("layer", "session"),
        dimension=body.get("dimension"),
        entity_id=body.get("entity_id"),
        valid_from=body.get("valid_from"),
        valid_until=body.get("valid_until"),
        importance=body.get("importance", 0.5),
        tags=body.get("tags") or [],
    )
    return JSONResponse(result)


async def memory_upload_endpoint(request: Request):
    """Memory upload REST endpoint - uploads and chunks file content."""
    from src.mcp.gateway import memory_add, DocumentLoader

    body = await request.json()

    file_name = body.get("file_name")
    file_content_base64 = body.get("file_content_base64")
    session_id = body.get("session_id")
    chunk_size = body.get("chunk_size", 1000)
    chunk_overlap = body.get("chunk_overlap", 200)

    if not isinstance(file_name, str) or not file_name.strip():
        return JSONResponse({"error": "file_name is required"}, status_code=400)
    if not isinstance(file_content_base64, str) or not file_content_base64.strip():
        return JSONResponse({"error": "file_content_base64 is required"}, status_code=400)
    if not isinstance(session_id, str) or not session_id.strip():
        return JSONResponse({"error": "session_id is required"}, status_code=400)

    try:
        if isinstance(file_content_base64, str) and "," in file_content_base64:
            file_content_base64 = file_content_base64.split(",", 1)[1]
        file_bytes = base64.b64decode(file_content_base64, validate=True)
    except Exception:
        return JSONResponse({"error": "invalid file_content_base64"}, status_code=400)

    try:
        text = DocumentLoader.load_file(io.BytesIO(file_bytes), file_name)
    except Exception as exc:
        return JSONResponse({"error": f"failed to parse file: {exc}"}, status_code=400)

    if not text.strip():
        return JSONResponse({"error": "file contains no readable text"}, status_code=400)

    chunk_size = int(chunk_size)
    chunk_overlap = int(chunk_overlap)
    if chunk_size <= 0:
        chunk_size = 1000
    if chunk_overlap < 0:
        chunk_overlap = 0
    if chunk_overlap >= chunk_size:
        chunk_overlap = max(chunk_size // 5, 0)

    try:
        try:
            from langchain_text_splitters import RecursiveCharacterTextSplitter
        except ImportError:
            from langchain.text_splitter import RecursiveCharacterTextSplitter

        splitter = RecursiveCharacterTextSplitter(
            chunk_size=chunk_size,
            chunk_overlap=chunk_overlap,
            length_function=len,
        )
        chunks = splitter.split_text(text)
    except Exception as exc:
        return JSONResponse({"error": f"failed to split file text: {exc}"}, status_code=500)

    if not chunks:
        return JSONResponse({"error": "file contains no indexable chunks"}, status_code=400)

    safe_filename = "".join([c for c in file_name if c.isalnum() or c in (" ", ".", "_")]).replace(" ", "_")
    if not safe_filename:
        safe_filename = "uploaded_file"

    tags = ["uploaded_material", f"filename:{safe_filename}", f"session:{session_id}"]
    memory_ids: List[str] = []

    for index, chunk in enumerate(chunks):
        chunk_content = chunk.strip()
        if not chunk_content:
            continue
        chunk_id = f"{session_id}_{safe_filename}_part_{index}"
        result = await memory_add(
            content=chunk_content,
            layer="session",
            dimension="context",
            entity_id=session_id,
            importance=0.6,
            tags=[*tags, f"chunk:{index}", f"chunk_id:{chunk_id}"],
        )
        memory_id = result.get("id") if isinstance(result, dict) else None
        if isinstance(memory_id, str):
            memory_ids.append(memory_id)

    if not memory_ids:
        return JSONResponse({"error": "failed to inject any file chunks"}, status_code=500)

    return JSONResponse({
        "status": "created",
        "file_name": file_name,
        "session_id": session_id,
        "chunks": len(memory_ids),
        "memory_ids": memory_ids,
    })


async def memory_temporal_endpoint(request: Request):
    """Memory temporal query REST endpoint."""
    from src.mcp.gateway import memory_get_temporal

    body = await request.json()
    result = await memory_get_temporal(
        entity_id=body.get("entity_id", ""),
        at_time=body.get("at_time"),
    )
    return JSONResponse(result)


__all__ = [
    "memory_search_endpoint",
    "memory_add_endpoint",
    "memory_upload_endpoint",
    "memory_temporal_endpoint",
]

File: src-ts/mcp/endpoints/memory.ts
/**
 * Memory REST Endpoints
 *
 * Memory-related HTTP endpoints for Desktop frontend.
 * Ported from src/mcp/endpoints/memory.py
 */

import type { HttpRequest, HttpResponse } from '../http-types';
import { jsonResponse, parseBody } from '../http-types';
import { memorySearch, memoryAdd, memoryGetTemporal } from '../services/memory';
import {
  normalizeProjectWorkspaceContext,
  projectWorkspaceToMemoryScope,
} from '../../project/workspace-model.js';
import { DocumentLoadError, DocumentLoader } from '../../services/document-loader';
import { recursiveCharacterSplit } from '../../ui/file-utils';
import { safeResolveWorkspaceRoot } from '../input-validation.js';

function resolveWorkspaceRoot(): string {
  return safeResolveWorkspaceRoot();
}

function readOptionalString(value: unknown): string | undefined {
  if (typeof value !== 'string') return undefined;
  const trimmed = value.trim();
  return trimmed ? trimmed : undefined;
}

function shouldUseFocusEntity(body: Record<string, unknown>): boolean {
  return body.use_focus_entity === true;
}

function resolveMemoryScope(
  body: Record<string, unknown>,
  options: { includeFocusEntity?: boolean } = {},
) {
  const workspace = normalizeProjectWorkspaceContext(body, {
    workspaceRoot: resolveWorkspaceRoot(),
  });
  return projectWorkspaceToMemoryScope(workspace, {
    includeFocusEntity: options.includeFocusEntity,
  });
}

/** POST /memory/search */
export async function memorySearchEndpoint(request: HttpRequest): Promise<HttpResponse> {
  const body = parseBody(request) as Record<string, unknown>;
  const explicitEntityId = readOptionalString(body.entity_id);
  const scope = resolveMemoryScope(body, {
    includeFocusEntity: explicitEntityId === undefined && shouldUseFocusEntity(body),
  });
  const result = await memorySearch({
    query: (body.query as string) ?? '',
    layer: body.layer as string | undefined,
    dimensions: body.dimensions as string[] | undefined,
    entityId: explicitEntityId ?? scope.entityId,
    userId: readOptionalString(body.user_id),
    projectId: readOptionalString(body.project_id) ?? scope.projectId,
    sessionId: readOptionalString(body.session_id) ?? scope.sessionId,
    atTime: readOptionalString(body.at_time),
    limit: (body.limit as number) ?? 10,
  });
  return jsonResponse(result);
}

/** POST /memory/add */
export async function memoryAddEndpoint(request: HttpRequest): Promise<HttpResponse> {
  const body = parseBody(request) as Record<string, unknown>;
  const explicitEntityId = readOptionalString(body.entity_id);
  const scope = resolveMemoryScope(body, {
    includeFocusEntity: explicitEntityId === undefined && shouldUseFocusEntity(body),
  });
  const result = await memoryAdd({
    content: (body.content as string) ?? '',
    layer: (body.layer as string) ?? 'session',
    dimension: body.dimension as string | undefined,
    entityId: explicitEntityId ?? scope.entityId,
    validFrom: readOptionalString(body.valid_from),
    validUntil: readOptionalString(body.valid_until),
    userId: readOptionalString(body.user_id),
    projectId: readOptionalString(body.project_id) ?? scope.projectId,
    sessionId: readOptionalString(body.session_id) ?? scope.sessionId,
    importance: (body.importance as number) ?? 0.5,
    tags: (body.tags as string[]) ?? [],
    source: readOptionalString(body.source),
    confidence: body.confidence as number | undefined,
  });
  return jsonResponse(result);
}

/** POST /memory/upload - uploads and chunks file content */
export async function memoryUploadEndpoint(request: HttpRequest): Promise<HttpResponse> {
  const body = parseBody(request) as Record<string, unknown>;
  const scope = resolveMemoryScope(body);

  const fileName = readOptionalString(body.file_name);
  const fileContentBase64 = readOptionalString(body.file_content_base64);
  const sessionId = readOptionalString(body.session_id) ?? scope.sessionId;
  const chunkSizeRaw = body.chunk_size as number | undefined;
  const chunkOverlapRaw = body.chunk_overlap as number | undefined;

  if (!fileName || !fileName.trim()) {
    return jsonResponse({ error: 'file_name is required' }, 400);
  }
  if (!fileContentBase64 || !fileContentBase64.trim()) {
    return jsonResponse({ error: 'file_content_base64 is required' }, 400);
  }
  if (!sessionId || !sessionId.trim()) {
    return jsonResponse({ error: 'session_id is required' }, 400);
  }

  let fileBuffer: Buffer;
  try {
    const normalized = fileContentBase64.includes(',')
      ? fileContentBase64.split(',', 2)[1]
      : fileContentBase64;
    const compact = normalized.replace(/\s+/g, '');
    const base64Pattern = /^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$/;
    if (!compact || compact.length % 4 !== 0 || !base64Pattern.test(compact)) {
      return jsonResponse({ error: 'invalid file_content_base64' }, 400);
    }
    fileBuffer = Buffer.from(normalized, 'base64');
    if (fileBuffer.length === 0) {
      return jsonResponse({ error: 'invalid file_content_base64' }, 400);
    }
    if (Buffer.from(fileBuffer.toString('base64'), 'base64').length !== fileBuffer.length) {
      return jsonResponse({ error: 'invalid file_content_base64' }, 400);
    }
  } catch {
    return jsonResponse({ error: 'invalid file_content_base64' }, 400);
  }

  let text: string;
  try {
    text = await DocumentLoader.loadFileAsync(fileBuffer, fileName);
  } catch (error) {
    if (error instanceof DocumentLoadError) {
      return jsonResponse(error.toResponseBody(fileName), 400);
    }
    return jsonResponse({ error: `failed to parse file: ${String(error)}` }, 400);
  }

  if (!text.trim()) {
    return jsonResponse({ error: 'file contains no readable text' }, 400);
  }

  let chunkSize = Number.isFinite(chunkSizeRaw) ? Math.trunc(chunkSizeRaw as number) : 1000;
  let chunkOverlap = Number.isFinite(chunkOverlapRaw) ? Math.trunc(chunkOverlapRaw as number) : 200;
  if (chunkSize <= 0) chunkSize = 1000;
  if (chunkOverlap < 0) chunkOverlap = 0;
  if (chunkOverlap >= chunkSize) {
    chunkOverlap = Math.floor(chunkSize / 5);
  }

  const chunks = recursiveCharacterSplit(text, chunkSize, chunkOverlap);
  if (!chunks.length) {
    return jsonResponse({ error: 'file contains no indexable chunks' }, 400);
  }

  const safeFilename = fileName
    .split('')
    .filter((char) => /[a-zA-Z0-9._ -]/.test(char))
    .join('')
    .trim()
    .replace(/\s+/g, '_') || 'uploaded_file';

  const memoryIds: string[] = [];
  const source = body.source as string | undefined;
  const confidence = body.confidence as number | undefined;
  for (let index = 0; index < chunks.length; index += 1) {
    const chunkContent = chunks[index].trim();
    if (!chunkContent) continue;
    const chunkId = `${sessionId}_${safeFilename}_part_${index}`;
    const result = await memoryAdd({
      content: chunkContent,
      layer: 'session',
      dimension: 'context',
      entityId: sessionId,
      importance: 0.6,
      tags: ['uploaded_material', `filename:${safeFilename}`, `session:${sessionId}`, `chunk:${index}`, `chunk_id:${chunkId}`],
      source,
      confidence,
    });

    if (result && typeof result === 'object' && typeof result['id'] === 'string') {
      memoryIds.push(result['id']);
    }
  }

  if (!memoryIds.length) {
    return jsonResponse({ error: 'failed to inject any file chunks' }, 500);
  }

  return jsonResponse({
    status: 'created',
    file_name: fileName,
    session_id: sessionId,
    chunks: memoryIds.length,
    memory_ids: memoryIds,
  });
}

/** POST /memory/temporal */
export async function memoryTemporalEndpoint(request: HttpRequest): Promise<HttpResponse> {
  const body = parseBody(request) as Record<string, unknown>;
  const scope = resolveMemoryScope(body);
  const entityId = readOptionalString(body.entity_id) ?? '';
  const result = await memoryGetTemporal(
    entityId,
    readOptionalString(body.at_time),
    {
      userId: readOptionalString(body.user_id),
      projectId: readOptionalString(body.project_id) ?? scope.projectId,
      sessionId: readOptionalString(body.session_id) ?? scope.sessionId,
    }
  );
  return jsonResponse(result);
}
